### PR TITLE
Improve ltp test failure error handling logic

### PR DESCRIPTION
**Pull request by sacsant**
LTP Test case on every run exits with failure even if there is no visible test case failure.

[124659.663797] Clock: inserting leap second 23:59:60 UTC
FAIL: No test failures detected, but LTP finished with 1 (1748.47 s)
RESULTS   : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 1748.52 s

This happens as the coded logic for failure detection does not cover all types
of failures.This patch rewrites the test failure detection code.
The new mechanism relies on ltp.log generated by LTP run. Since ltp.log
contains information about test failures marked as BROK (in addition to TFAIL,
TCONF, TBROK) there are very little chances of missing any type of failures.

Signed-off-by: Sachin Sant <sachinp@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by sacsant** [comment]
# avocado run generic/ltp.py 
JOB ID     : 30fa7a1bf5416cd245afe19295ac6740ca2f98fa
JOB LOG    : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-12T09.43-30fa7a1/job.log
 (1/1) generic/ltp.py:ltp.test: - FAIL: LTP tests failed: ['msgctl04', 'dirtyc0w'] (872.85 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 872.91 s
JOB HTML   : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-12T09.43-30fa7a1/results.html

In this case eg dirtyc0w test is not marked as FAIL in the debug log. Following messages can be seen in the log

2019-04-12 09:53:22,306 process          L0451 DEBUG| [stdout] tst_test.c:1085: INFO: Timeout per run is 0h 05m 00s
2019-04-12 09:53:22,306 process          L0451 DEBUG| [stdout] dirtyc0w.c:81: BROK: execlp(dirtyc0w_child, dirtyc0w_child, ...) failed: EACCES
2019-04-12 09:53:32,288 process          L0451 DEBUG| [stdout] tst_checkpoint.c:144: BROK: dirtyc0w.c:84: tst_checkpoint_wait(0, 10000): ETIMEDOUT
2019-04-12 09:53:32,288 process          L0451 DEBUG| [stdout]
2019-04-12 09:53:32,288 process          L0451 DEBUG| [stdout] Summary:
2019-04-12 09:53:32,288 process          L0451 DEBUG| [stdout] passed   0
2019-04-12 09:53:32,288 process          L0451 DEBUG| [stdout] failed   0
2019-04-12 09:53:32,289 process          L0451 DEBUG| [stdout] skipped  0
2019-04-12 09:53:32,289 process          L0451 DEBUG| [stdout] warnings 0

File: generic/ltp.py [hunk 1]
```diff
@@ -58,29 +60,26 @@ def test(self):
         args = self.params.get('args', default='')
         logfile = os.path.join(self.logdir, 'ltp.log')
         failcmdfile = os.path.join(self.logdir, 'failcmdfile')
+        failed_tests = False
```
Comment (harish-24) [inline generic/ltp.py]: Can we replace `values` list as `failed_tests` list and avoid using `failed_tests` bool?

File: generic/ltp.py [hunk 2]
```diff
@@ -58,29 +60,26 @@ def test(self):
...
+                    failed_tests = True
+                    value = re.split(r'\s+', line)
+                    self.values.append(value[0])
 
         if failed_tests:
-            self.fail("LTP tests failed: %s" % ", ".join(failed_tests))
-        elif result.exit_status != 0:
-            self.fail("No test failures detected, but LTP finished with %s"
-                      % (result.exit_status))
+            self.fail("LTP tests failed: %s" % self.values)
```
Comment (harish-24) [inline generic/ltp.py]: we can directly check with the empty list here also.

Comment (sacsant) [reply generic/ltp.py]: V2 patch on its way

**Comment by sacsant** [comment]
Run logs from few test runs:

LTP invoked with -f syscalls

# avocado run generic/ltp.py 
JOB ID     : b92ca62a02cbb05fc6b4afb98eeb34d5b68959b6
JOB LOG    : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-15T06.28-b92ca62/job.log
 (1/1) generic/ltp.py:ltp.test: |
Broadcast message from systemd-journald@ltczep10-lp4.aus.stglabs.ibm.com (Mon 2019-04-15 06:40:42 CDT):

syslogtst[38958]: syslogtst: mail emerg test.

|
Broadcast message from systemd-journald@ltczep10-lp4.aus.stglabs.ibm.com (Mon 2019-04-15 06:42:07 CDT):

syslogtst[39245]: syslogtst: emergency log

FAIL: LTP tests failed: ['msgctl04', 'dirtyc0w'] (871.45 s)
RESULTS    : PASS 0 | ERROR 0 | FAIL 1 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 871.51 s
JOB HTML   : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-15T06.28-b92ca62/results.html

 LTP invoked with -f math
 
# avocado run generic/ltp.py 
JOB ID     : 8d03b1540c874fa9f93316f77d95c022d0a68756
JOB LOG    : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-15T06.44-8d03b15/job.log
 (1/1) generic/ltp.py:ltp.test: PASS (115.46 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 115.53 s
JOB HTML   : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-15T06.44-8d03b15/results.html

ltp.log contains following: 
Testcase                                           Result     Exit Value
--------                                           ------     ----------
abs01                                              PASS       0
atof01                                             PASS       0
float_bessel                                       PASS       0
float_exp_log                                      PASS       0
float_iperb                                        PASS       0
float_power                                        PASS       0
float_trigo                                        PASS       0
fptest01                                           PASS       0
fptest02                                           PASS       0
nextafter01                                        PASS       0

-----------------------------------------------
Total Tests: 10
Total Skipped Tests: 0
Total Failures: 0


LTP invoked with -f power_management_tests
# 
# avocado run generic/ltp.py 
JOB ID     : d8bd8eda243678ebd25c67e16749ea82ff50631c
JOB LOG    : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-15T06.47-d8bd8ed/job.log
 (1/1) generic/ltp.py:ltp.test: PASS (108.43 s)
RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 108.48 s
JOB HTML   : /home/sachin/avocado-fvt-wrapper/results/job-2019-04-15T06.47-d8bd8ed/results.html
# 

ltp.log contains following :
Testcase                                           Result     Exit Value
--------                                           ------     ----------
runpwtests01                                       CONF       32
runpwtests02                                       CONF       32
runpwtests03                                       CONF       32
runpwtests04                                       CONF       32
runpwtests06                                       CONF       32

-----------------------------------------------
Total Tests: 5
Total Skipped Tests: 5
Total Failures: 0
Kernel Version: 5.1.0-rc1
Machine Architecture: ppc64le

File: generic/ltp.py [hunk 3]
```diff
@@ -65,22 +67,17 @@ def test(self):
...
+                if 'FAIL' in line:
+                    value = re.split(r'\s+', line)
+                    self.failed_tests.append(value[0])
 
-        if failed_tests:
-            self.fail("LTP tests failed: %s" % ", ".join(failed_tests))
-        elif result.exit_status != 0:
-            self.fail("No test failures detected, but LTP finished with %s"
-                      % (result.exit_status))
+        if len(self.failed_tests) != 0:
```
Comment (harish-24) [inline generic/ltp.py]: Minor suggestion: `if self.failed_tests:` would suffice here and also abides pylint standard.

Comment (sacsant) [reply generic/ltp.py]: I think both forms are valid .. but yeah this can be changed to just self.failed_tests.  Will send updated patch

**Comment by PraveenPenguin** [comment]
@sacsant thanks for patch , @harish-24 thanks for ACK

**Pull request merged**